content/Interfaces/Interfaces.q.md

```diff
@@ -0,0 +1,2450 @@
+
+ - Abstract function `Person::display()` cannot be declared private on line number 46
+
+ - Access type for interface method `LifeCycle::species()` must be omitted on line number 4
+
+:::
+
+
+/// type=CR, answer=[tests/Interfaces/CorrectMultipleErrorsTest.php], filename=[LifeCycle.php,Person.php,Student.php]
+
+Correct the code so that it outputs the string `John is an adolescent human being.`.
+
+```php
+<?php
+interface LifeCycle
+{
+  protected function species();
+  public function stage();
+}
+?>
+```
+
+```php
+<?php
+abstract class Person
+{
+    protected $name;
+    private $age;
+	
+    public function __construct($name, $age)
+    {
+        $this->name = $name;
+        if ($this->checkAge($age)) {
+            $this->age = $age;
+        }
+    }
+    
+    public function setAge($newAge)
+    {
+        if ($this->checkAge($newAge)) {
+            $this->age = $newAge;
+        }
+    }
+
+    public function getAge()
+    {
+        return $this->age;
+    }
+
+    public function getName()
+    {
+        return $this->name;
+    }
+
+    private function checkAge($age)
+    {
+        if ($age > 0) {
+            return true;
+        }
+        return false;
+    }
+
+    abstract private function display();
+}
+?>
+```
+
+```php
+<?php 
+class Student Person LifeCycle
+{
+    public function species()
+    {
+        return "human";
+    }
+	
+    public function stage()
+    {
+        return "adolescent";
+    }
+	
+    public function display()
+    {
+        echo $this->name . " is an " . $this->stage() . " " . $this->species() . " being.";  
+    }
+}
+
+$studObject = new Person("John", 15);
+$studObject->display();
+?>
+```
+
++++
+
+
++++
+
+### Part 4: Practice
+
+/// type=CR, answer=[tests/Interfaces/CreateClassThatImplementsAnInterfaceTest.php], filename=[LifeCycle.php,Animal.php,Mammal.php]
+
+Given the `Animal` class and `LifeCycle` interface, write a program that uses an `implements` keyword to implement an interface. In the `Mammal.php` tab, use a `class` keyword to declare a class named `Mammal` that `extends` the `Animal` class and `implements` the `LifeCycle` interface. Within the curly braces `{}`, add an implementation or method body to all abstract methods of the `LifeCycle` interface. Inside the `stage()` method body, add a statement `return "adult";`. Inside the `species()` method body, add a statement that returns the string `animal`. Then, add another `public` method `display()` that overrides the abstract `display()` method of the `Animal` class. Inside the `display()` method body, add an `echo` statement to display the string `"The " . parent::getType() . " is an " . $this->stage() . " " . $this->species() . "."`. After the class declaration, add a statement that creates the `$petMammal` object an instance of the `Mammal` class passing the arguments `Cat` and `3`. Then, add another statement that calls the `display()` method of the `$petMammal` object. Run the program to view the output.
```

Comment:
What if: In the `Mammal.php` tab, use the `class` keyword to declare a class named `Mammal`.....

Reply:
This is already discussed before that we use the article a in part 4 and part 2. Hope you can still recall. Please check. Thanks.